### PR TITLE
feat(Logical Context): The equivalent to log4net.ThreadLogicalContext

### PR DESCRIPTION
**Pull request by bfcamara**
Adds Support for the equivalent to log4net.ThreadLogicalContext

The MappedDiagnosticsContext stores the dictionary in the thread locall storage, which it does not flow in async points. For example, if I am using the async/await pattern in my web application, and I want to log a specific property (being set in the begining of the request processing), in all my log statements, if I use the MappedDiagnosticsContext, I will lose the property in logs as soon the request jump to another thread.

This PR will allows us to use a class,MappedDiagnosticsLogicalContext, which uses the CallContext to store the items. A new layout renderer is included as well (${mdlc}).

### CODE REVIEW

**Comment by BrutalCode** [comment]
I read code and found some difference between `MappedDiagnosticsContext` and `MappedDiagnosticsLogicalContext`.

`MappedDiagnosticsLogicalContext` has only one `Set` method that accepts string value. Also it stores values in `ConcurrentDictionary` that TVal is **string**.

But `MappedDiagnosticsContext` has two `Set` methods that accepts string and object values. Also it stores values in `Dictionary` that TVal is **object**.

In other words, `MappedDiagnosticsLogicalContext` does not store object type values while `MappedDiagnosticsContext` do.
 
Why does this difference exist? Is there any problem with using object value in `MappedDiagnosticsLogicalContext`.

I read this issue to find why, but I could not see any reason. Maybe it is so trivial, but I want to ask.

**Comment by bfcamara** [comment]
@BrutalCode 

The support to storing objects in `MappedDiagnosticsContext` is recent, since last June. You can check the commit here https://github.com/NLog/NLog/commit/db66454607d27c15beaa9976c4ac9fcaa6bc7b2f. The `MappedDiagnosticLogicalContext`was added 2 years ago. 

I think there is no reason to not use the same approach in  `MappedDiagnosticLogicalContext` as long as the objects being added to the concurrent dictionary are serializable. The `MappedDiagnosticLogicalContext` was mainly written with await/async in mind,  in order to have values flowing between threads. By that time I just needed to have string values.

**Comment by BrutalCode** [comment]
@bfcamara Thank you for response! You are right. I can see the reason.

I am working for MDLC wiki page, and to do this, I am reading the source code both of them (MappedDiagnosticsLogicalContext and MappedDiagnosticsContext). 

So I just needed to ask you to write wiki page correctly.

Thank you for your efforts!

By the way, I need to be sure for its wiki page before save it. Could you help me?

**Comment by 304NotModified** [comment]
@brutalcode. If the page is not in the index yet, there is no problem to save it. 

And see also https://github.com/NLog/NLog/issues/883

**Comment by Fresa** [comment]
The implementation does have some quirky behavior when you are using a mutable dictionary. You should change to an immutable dictionary since CallContext uses shallow copy on write on thread/task delegation.

Mr Cleary explains it well in his blog: http://blog.stephencleary.com/2013/04/implicit-async-context-asynclocal.html

**Comment by 304NotModified** [comment]
Thanks for the info. 

Note: this repository is deprecated.  This feature has been moved to NLog core.

**Comment by Fresa** [comment]
I noticed that, so I've created an issue in the NLog repo instead.

**Comment by 304NotModified** [comment]
Thanks!
